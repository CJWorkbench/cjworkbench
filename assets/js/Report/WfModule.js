import React from 'react'
import PropTypes from 'prop-types'
import propTypes from 'prop-types'
import OutputIframe from '../OutputIframe'

const WfModule = React.memo(function WfModule ({ workflowId, isPublic, id, deltaId }) {
  // Assume the chart itself has a title -- no need to write a title here.
  //
  // But we'll link back to the Workflow, so users can edit a chart.
  return (
    <figure>
      <OutputIframe
        visible
        workflowId={workflowId}
        isPublic={isPublic}
        wfModuleId={id}
        deltaId={deltaId}
      />
      <figcaption>
        <a href='#' onClick={/*navigateToWfModule*/null}>Jump to Workflow</a>
      </figcaption>
    </figure>
  )
})
WfModule.propTypes = {
  workflowId: PropTypes.number.isRequired,
  isPublic: PropTypes.bool.isRequired,
  ...propTypes.WfModule
}
export default WfModule
