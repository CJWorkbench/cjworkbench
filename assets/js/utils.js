// ---- Utilities ---
import * as Cookies from 'js-cookie'
import { fromByteArray as base64Encode } from 'base64-js'
import { t } from '@lingui/macro'

export function goToUrl (url) {
  window.location.href = url
}

// Current CSRF token
export const csrfToken = Cookies.get('csrftoken')

// Gets the letter coordinate of a column from its index within the column names array
export function idxToLetter (idx) {
  var letters = ''
  var cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode(cidx % 26 + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}

// Log to Intercom, if installed
export function logUserEvent (name, metadata) {
  if (!window.APP_ID) return

  // If we're in a lesson, drop the event.
  if (window.initState && window.initState.lessonData) return

  window.Intercom('trackEvent', name, metadata)
}

export function timeDifference (start, end, i18n) {
  const ms = new Date(end) - new Date(start)
  const seconds = Math.floor(ms / 1000)
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)
  const days = Math.floor(hours / 24)
  const years = Math.floor(days / 365.25)

  if (years > 0) {
    return i18n._(t('js.util.timeDifference.ago.years')`${years}y ago`)
  } else if (days > 0) {
    return i18n._(t('js.util.timeDifference.ago.days')`${days}d ago`)
  } else if (hours > 0) {
    return i18n._(t('js.util.timeDifference.ago.hours')`${hours}h ago`)
  } else if (minutes > 0) {
    return i18n._(t('js.util.timeDifference.ago.minutes')`${minutes}m ago`)
  } else if (seconds > 0) {
    return i18n._(t('js.util.timeDifference.ago.seconds')`${seconds}s ago`)
  } else {
    return i18n._(t('js.util.timeDifference.now')`just now`)
  }
}

export function escapeHtml (str) {
  str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;')

  return str
}

/**
 * Scroll `containerEl` vertically by the smallest amount possible to put
 * `el` in view.
 *
 * Set the container CSS `scroll-behavior: smooth` to make this transition
 * look nice.
 *
 * @param el Element we want to focus
 * @param containerEl Element we'll set scrollTop on so it contains el
 * @param marginTop Minimum number of pixels between containerEl.top and el.top
 * @param marginBottom Minimum number of pixels between containerEl.bottom and el.bottom
 */
export function scrollTo (el, containerEl, marginTop, marginBottom) {
  if (marginTop === undefined) marginTop = 10
  if (marginBottom === undefined) marginBottom = 10

  const elRect = el.getBoundingClientRect()
  const containerRect = containerEl.getBoundingClientRect()

  // Calculate dy: how much do we need to add to containerEl.scrollTop to
  // make el go where we want it to go?
  let dy = 0 // Prefer not to scroll
  if (elRect.bottom + marginBottom > containerRect.bottom) {
    // Scroll down if we need to.
    // We do this before up because some els might be taller than containerEl,
    // and in that case we want scroll-up to override scroll-down (because the
    // top of el is more important than the bottom).
    dy += (elRect.bottom + marginBottom - containerRect.bottom)
  }
  if (dy + elRect.top - marginTop < containerRect.top) {
    // Scroll up if we need to.
    dy -= (containerRect.top - (elRect.top - marginTop))
  }

  containerEl.scrollTop += dy
}

/**
 * Build a unique slug with the given prefix.
 *
 * The slug uses randomly-generated bytes: 9 bytes, base64-encoded with
 * '+' and '/' changed to '-' and '_' (to fit a Django "slug" field).
 *
 * Django 'slug' regex: r'^[-a-zA-Z0-9_]+\Z'
 *
 * According to https://en.wikipedia.org/wiki/Birthday_attack, 9 bytes -- 72
 * bits -- means if clients call this 97,000 times, they have a <10^-12 chance
 * of generating a collision within a given Workflow. This should protect even
 * the most click-happy user. (6 bytes would allow 380 before the chance grows
 * higher, and 380 steps is possible.)
 *
 * (Why 9 bytes and not 8? Because base64-encoding gives the same string
 * length. 9 bytes encoded as base64 gives 12 bytes.)
 */
export function generateSlug (prefix) {
  const bytes = new Uint8Array(9)
  window.crypto.getRandomValues(bytes)
  return prefix + (
    base64Encode(bytes)
      .replace(/\+/g, '-')
      .replace(new RegExp('/', 'g'), '_')
  )
}
