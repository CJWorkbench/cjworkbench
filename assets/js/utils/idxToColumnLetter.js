/**
 * Return 'A' for 0, 'B' for 1, 'AA' for 26, and so on.
 */
export default function idxToColumnLetter (idx) {
  let letters = ''
  let cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode((cidx % 26) + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}
