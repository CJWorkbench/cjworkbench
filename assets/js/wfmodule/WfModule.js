// UI for a single module within a workflow

import React from 'react'
import WfParameter from '../WfParameter'
import WfModuleContextMenu from '../WfModuleContextMenu'
import EditableNotes from '../EditableNotes'
import StatusBar from './StatusBar'
import StatusLine from './StatusLine'
import {
  store,
  setWfModuleCollapsedAction,
  updateWfModuleAction,
  clearNotificationsAction,
  setSelectedWfModuleAction
} from '../workflow-reducer'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import lessonSelector from '../lessons/lessonSelector'


// ---- WfModule ----
export class WfModule extends React.PureComponent {
  constructor(props) {
    super(props);

    this.click = this.click.bind(this);
    this.changeParam = this.changeParam.bind(this);
    this.setParamText = this.setParamText.bind(this);
    this.getParamText = this.getParamText.bind(this);
    this.getParamMenuItems = this.getParamMenuItems.bind(this);
    this.removeModule = this.removeModule.bind(this);
    this.showNotes = this.showNotes.bind(this);
    this.hideNotes = this.hideNotes.bind(this);
    this.setNotifications = this.setNotifications.bind(this);
    this.setClickNotification = this.setClickNotification.bind(this);
    this.onClickNotification = this.onClickNotification.bind(this);
    this.setModuleRef = this.setModuleRef.bind(this);
    this.moduleRef = null;

    this.state = {
      isCollapsed: this.props.wfModule.is_collapsed,
      showNotes:  ( this.props.wfModule.notes
                    && (this.props.wfModule.notes != "")
                    && (this.props.wfModule.notes != "Write notes here")
                  ),  // only show on load if a note exists & not default text
      showEditableNotes: false,             // do not display in edit state on initial load
      notifications: this.props.wfModule.notifications,
      notification_count: this.props.wfModule.notification_count,
      isDragging: false,
    }
  }

  // pass a function to all wf_parameters to allow them to overload
  // the function that runs when a user clicks on the notification
  // icon
  setClickNotification(cb) {
    this.clickNotification = cb;
  }

  clickNotification() {
    return false;
  }

  onClickNotification() {
    store.dispatch(clearNotificationsAction(this.props.wfModule.id));
    this.clickNotification();
  }

  componentWillReceiveProps(newProps) {
    if (newProps.wfModule.is_collapsed !== this.state.isCollapsed) {
      this.setState({
        isCollapsed: newProps.wfModule.is_collapsed
      })
    }
  }

  // Scroll when we create a new wfmodule
  componentDidMount() {
    if (this.props.selected) {
      this.props.focusModule(this.moduleRef);
    }
  }

  // We become the selected module on any click
  click(e) {
    store.dispatch(setSelectedWfModuleAction(this.props.wfModule.id));
  }

  changeParam(id, payload) {
    this.props.changeParam(id, payload)
  }

  onDragStart = (ev) => {
    if (ev.target.tagName in {
      'button': null,
      'input': null,
      'select': null,
      'textarea': null,
      'label': null,
    }) {
      // Don't drag when user selects text
      ev.preventDefault()
      return
    }

    const dragObject = {
      type: 'WfModule',
      index: this.props.index,
      id: this.props.wfModule.id,
    }
    ev.dataTransfer.setData('application/json', JSON.stringify(dragObject))
    ev.dataTransfer.effectAllowed = 'move'
    ev.dataTransfer.dropEffect = 'move'
    this.props.onDragStart(dragObject)

    this.setState({
      isDragging: true,
    })
  }

  onDragEnd = (ev) => {
    this.props.onDragEnd()

    this.setState({
      isDragging: false,
    })
  }

  // These functions allow parameters to access each others value (text params only)
  // Used e.g. for custom UI elements to save/restore their state from hidden parameters
  // Suppresses reassignment of the same text, which can be important to avoid endless notification loops
  setParamText(paramIdName, text) {
    var p = this.props.wfModule.parameter_vals.find( p => p.parameter_spec.id_name == paramIdName );
    if (p && text != p.string) {
      this.props.changeParam(p.id, { value: text })
    }
  }

  getParamText(paramIdName) {
    var p = this.props.wfModule.parameter_vals.find( p => p.parameter_spec.id_name == paramIdName );
    if (p) {
      return p.value;
    }
  }

  getParamId = (paramIdName) => {
    var p = this.props.wfModule.parameter_vals.find( p => p.parameter_spec.id_name == paramIdName );
    if (p) {
      return p.id;
    }
  }

  getParamMenuItems(paramIdName) {
    var p = this.props.wfModule.parameter_vals.find(p => p.parameter_spec.id_name == paramIdName);
    if(p) {
      if(p.menu_items) {
        return p.menu_items.split('|').map(s => s.trim());
      }
    }
    return [];
  }

  removeModule(e) {
    this.props.removeModule(this.props.wfModule.id);
  }

  // Optimistically updates the state, and then sends the new state to the server,
  // where it's persisted across sessions and through time.
  setCollapsed(isCollapsed) {
    this.setState({
      isCollapsed,
    })

    store.dispatch(setWfModuleCollapsedAction(
      this.props.wfModule.id,
      isCollapsed,
      this.props.isReadOnly
    ))
  }

  collapse = () => {
    this.setCollapsed(true)
  }

  expand = () => {
    this.setCollapsed(false)
  }

  // when Notes icon is clicked, show notes and start in editable state if not read-only
  showNotes(e) {
    e.stopPropagation();
    this.setState({ showNotes: true,  showEditableNotes: !this.props.isReadOnly });
  }

  hideNotes() {
    this.setState({showNotes: false});
  }

  setNotifications() {
    store.dispatch(
      updateWfModuleAction(
        this.props.wfModule.id,
        { notifications: !this.props.wfModule.notifications }
    ));
  }

  setModuleRef(ref) {
    this.moduleRef = ref;
  }

  render() {
    let wfModule = this.props.wfModule;
    let module = wfModule.module_version.module;

    var updateSettings = {
      lastUpdateCheck:  wfModule.last_update_check,
      autoUpdateData:   wfModule.auto_update_data,
      updateInterval:   wfModule.update_interval,
      updateUnits:      wfModule.update_units
    };

    // Each parameter gets a WfParameter
    var paramdivs = wfModule.parameter_vals.map((ps, i) => {
        return (<WfParameter
          api={this.props.api}
          moduleName={module.name}
          isReadOnly={this.props.isReadOnly}
          key={i}
          p={ps}
          changeParam={this.changeParam}
          wf_module_id={wfModule.id}
          revision={this.props.revision}
          updateSettings={updateSettings}
          getParamId={this.getParamId}
          getParamText={this.getParamText}
          getParamMenuItems={this.getParamMenuItems}
          setParamText={this.setParamText}
          setClickNotification={this.setClickNotification}
          notifications={wfModule.notifications}
          loggedInUser={this.props.user}
        />)
      });

    var notes;
    var value = ( wfModule.notes && (wfModule.notes != "") )
      ? wfModule.notes
      : "Write notes here";

    if (this.state.showNotes)
      notes = <div className='module-notes'>
                <EditableNotes
                  api={this.props.api}
                  isReadOnly={this.props.isReadOnly}
                  value={value}
                  hideNotes={ () => this.hideNotes() }
                  wfModuleId={wfModule.id}
                  startFocused={this.state.showEditableNotes}
                />
              </div>;

    let helpIcon;
    if (!this.props.isReadOnly) {
      helpIcon = (
        <a title='Help for this module' className='help-button' href={module.help_url} target='_blank'>
          <i className='icon-help'></i>
        </a>
      );
    }

    var notesIcon;
    if (!this.state.showNotes && !this.props.isReadOnly)
      notesIcon = <button title="Edit Note" className={'btn edit-note' + (this.props.isLessonHighlightNotes ? ' lesson-highlight' : '')} onClick={this.showNotes}>
                    <i className='icon-note'></i>
                  </button>;

    var contextMenu;
    if(!this.props.isReadOnly)
      contextMenu = <WfModuleContextMenu
          removeModule={ () => this.removeModule() }
          stopProp={(e) => e.stopPropagation()}
          id={wfModule.id}

        />;


    // Set opacity to 0/1 instead of just not rendering these elements, so that any children that these
    // buttons create (e.g. export dialog) are still visible. Can't use display: none as we need display: flex
    // Fixes https://www.pivotaltracker.com/story/show/154033690
    const contextBtns =
        <div className='context-buttons'>
          {wfModule.notifications &&
          <button
            className={'notification-badge' + (wfModule.notification_count > 0 ? ' active t-f-blue' : '' )}
            onClick={this.onClickNotification}
            >
            <i className="icon-notification"></i>
            { wfModule.notification_count > 0 && <span className="count">{wfModule.notification_count}</span> }
          </button>
          }
          {helpIcon}
          {notesIcon}
          {contextMenu}
        </div>

    const moduleIcon = 'icon-' + module.icon + ' WFmodule-icon mr-2';

    // Putting it all together: name, status, parameters, output
    return (
      <div onClick={this.click} className={'wf-module' + (this.props.isLessonHighlight ? ' lesson-highlight' : '') + (this.state.isCollapsed ? ' collapsed' : ' expanded')} data-module-name={module.name}>
        {notes}
        <div className={'wf-card '+ (this.state.isDragging ? 'dragging ' : '')} ref={this.setModuleRef} draggable={!this.props.isReadOnly} onDragStart={this.onDragStart} onDragEnd={this.onDragEnd}>

          <div>
            <div className='output-bar-container'>
              <StatusBar status={wfModule.status} isSelected={this.props.selected} isDragging={this.state.isDragging}/>
            </div>
            <div className='module-content'>
              <div className='module-card-header'>
                <WfModuleCollapseButton
                  isCollapsed={this.state.isCollapsed}
                  isLessonHighlight={this.props.isLessonHighlightCollapse}
                  onCollapse={this.collapse}
                  onExpand={this.expand}
                  />
                <i className={moduleIcon}></i>
                <div className='module-name'>{module.name}</div>
                {contextBtns}
              </div>
              {/* --- Error message --- */}
              <StatusLine status={wfModule.status} error_msg={wfModule.error_msg} />
              <div className='module-card-params'>
                {paramdivs}
              </div>
            </div>
            <div className={
              'drop-alert ' +
              ( (this.props.dragItemType === 'notification' && this.props.canDrop && this.props.dragItem) ? 'active ' : '' ) +
              ( (this.props.dragItemType === 'notification' && this.props.canDrop && this.props.isOver) ? 'over ' : '')
              } />

          </div>
        </div>
      </div>
    ) || null;
  }
}
WfModule.propTypes = {
  isReadOnly:         PropTypes.bool.isRequired,
  index:              PropTypes.number.isRequired,
  wfModule:           PropTypes.object,
  revison:            PropTypes.number,
  selected:           PropTypes.bool,
  changeParam:        PropTypes.func,
  removeModule:       PropTypes.func,
  api:                PropTypes.object.isRequired,
  onDragStart:        PropTypes.func.isRequired, // func({ type:'WfModule',id,index }) => undefined
  onDragEnd:          PropTypes.func.isRequired, // func() => undefined
  focusModule:        PropTypes.func,
  isLessonHighlight: PropTypes.bool.isRequired,
  isLessonHighlightNotes: PropTypes.bool.isRequired,
  isLessonHighlightCollapse: PropTypes.bool.isRequired,
}

class WfModuleCollapseButton extends React.PureComponent {
  static propTypes = {
    isCollapsed: PropTypes.bool.isRequired,
    isLessonHighlight: PropTypes.bool.isRequired,
    onCollapse: PropTypes.func.isRequired, // func() => undefined
    onExpand: PropTypes.func.isRequired, // func() => undefined
  }

  render() {
    const { isCollapsed, isLessonHighlight, onCollapse, onExpand } = this.props

    const iconClass = isCollapsed ? 'icon-sort-right' : 'icon-sort-down'
    const onClick = isCollapsed ? onExpand : onCollapse
    const name = isCollapsed ? 'expand module' : 'collapse module'
    const lessonHighlightClass = isLessonHighlight ? 'lesson-highlight' : ''
    return (
      <button name={name} className="wf-module-collapse" onClick={onClick}>
        <i className={`context-collapse-button ${iconClass} ${lessonHighlightClass}`}></i>
      </button>
    )
  }
}

function propsToModuleName(props) {
  return (
    props.wfModule
    && props.wfModule.module_version
    && props.wfModule.module_version.module
    && props.wfModule.module_version.module.name
    || ''
  )
}

function mapStateToProps(state, ownProps) {
  const { testHighlight } = lessonSelector(state)
  const moduleName = propsToModuleName(ownProps)
  return {
    isLessonHighlight: testHighlight({ type: 'WfModule', moduleName }),
    isLessonHighlightCollapse: testHighlight({ type: 'WfModuleContextButton', button: 'collapse', moduleName }),
    isLessonHighlightNotes: testHighlight({ type: 'WfModuleContextButton', button: 'notes', moduleName }),
  }
}

// TODO replace "store.dispatch" with mapDispatchToProps()

export default connect(
  mapStateToProps,
)(WfModule)
