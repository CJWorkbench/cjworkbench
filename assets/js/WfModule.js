// UI for a single module within a workflow

import React, { PropTypes } from 'react'
import ChartParameter from './Chart'
import { store, wfModuleStatusAction } from './workflow-reducer'
import { csrfToken } from './utils'

// Libraries to provide a collapsable table view
import { Collapse, Button, CardBlock, Card } from 'reactstrap';
import ReactDataGrid from 'react-data-grid';


// ---- WfParameter - a single editable parameter ----

class WfParameter extends React.Component {

  constructor(props) {
    super(props);

    this.type = this.props.p.parameter_spec.type;
    this.name = this.props.p.parameter_spec.name;

    this.keyPress = this.keyPress.bind(this);
    this.blur = this.blur.bind(this);
    this.click = this.click.bind(this);
  }

  paramChanged(e) {
    // console.log("PARAM CHANGED");
    var newVal = {};
    newVal[this.type] = e.target.value;
    this.props.changeParam(this.props.p.id, newVal);
  }

  // Save value (and re-render) when user presses enter or we lose focus
  // Applies only to non-text fields
  keyPress(e) {
    if (this.type != 'text' && e.key == 'Enter') {
        this.paramChanged(e);
        e.preventDefault();       // eat the Enter so it doesn't get in out input field
    }
  }

  blur(e) {
    this.paramChanged(e);
  }

  // Send event to server for button click
  click(e) {
    if (this.type == 'button') {
      var url = '/api/parameters/' + this.props.p.id + '/event';
      var eventData = {'type': 'click'};
      fetch(url, {
        method: 'post',
        credentials: 'include',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'X-CSRFToken': csrfToken
        },
        body: JSON.stringify(eventData)
      }).then(response => {
        if (!response.ok)
          store.dispatch(wfModuleStatusAction(this.props.wf_module_id, 'error', response.statusText))
      });
    }
    if (this.type == 'checkbox') {
        var newVal = {};
        newVal[this.type] = e.target.checked;
        this.props.changeParam(this.props.p.id, newVal);
    }
  }

  render() {
    if (!this.props.p.visible) {
      return false; // nothing to see here
    }

    switch (this.type) {
      case 'string':
        return (
          <div>
            <div>{this.name}:</div>
            <textarea className='wfmoduleStringInput' rows='1' defaultValue={this.props.p.string} onBlur={this.blur} onKeyPress={this.keyPress} />
          </div>
        );

      case 'number':
        return (
          <div>
            <div>{this.name}:</div>
            <textarea className='wfmoduleNumberInput' rows='1' defaultValue={this.props.p.number} onBlur={this.blur} onKeyPress={this.keyPress} />
          </div>
        );

      case 'text':
        return (
          <div>
            <div>{this.name}:</div>
            <textarea className='wfmoduleTextInput' rows='4' defaultValue={this.props.p.text} onBlur={this.blur} onKeyPress={this.keyPress} />
          </div>
        );

      case 'button':
        return (
          <div>
            <button className='btn btn-secondary' onClick={this.click}>{this.name}</button>
          </div>
        );

      case 'checkbox':
        return (
            <div>
                <label className='mr-1'>{this.name}:</label>
                <input type="checkbox" checked={this.props.p.checkbox} onChange={this.click}></input>
            </div>
        );

      case 'custom':

        // Load and save chart state, image to hidden parameters
        var loadState = ( () => this.props.getParamText('chartstate') );
        var saveState = ( state => this.props.setParamText('chartstate', state) );

        var saveImageDataURI = ( state => this.props.setParamText('chart', state) );

        return (
          <div>
            <a href={'/public/paramdata/live/' + this.props.p.id + '.png'}>PNG</a>
            <ChartParameter
              wf_module_id={this.props.wf_module_id}
              revision={this.props.revision}
              saveState={saveState}
              loadState={loadState}
              saveImageDataURI={saveImageDataURI}
            />
          </div>
        );

      default:
        return null;  // unrecognized parameter type
    }
  }
}

WfParameter.propTypes = {
  p:                React.PropTypes.object,
  wf_module_id:     React.PropTypes.number,
	revision:         React.PropTypes.number,
  changeParam:      React.PropTypes.func,
	getParamText:     React.PropTypes.func,
	setParamText:     React.PropTypes.func,
};


// ---- StatusLight ----
// Ready, Busy, or Error
class StatusLight extends React.Component {
  render() {
    return <div className={this.props.status + '-light'}></div>
  }
}

// ---- StatusLine ----
// Display error message, if any
class StatusLine extends React.Component {
  render() {
    if (this.props.status == 'error') {
      return <div className='wfModuleErrorMsg'>{this.props.error_msg}</div>
    } else if (this.props.status == 'busy') {
      return <div className='wfModuleErrorMsg'>Working...</div>
    } else {
      return false
    }
  }
}

// ---- CollapseSection ---
// Higher-order component that does a classic twirly collapse, inside a rounded border

function CollapseSection(WrappedComponent, title, startOpen ) {
  return class extends React.Component {
    constructor(props) {
      super(props);
      this.state = {isOpen: startOpen};           // componentDidMount will trigger first load
      this.toggle = this.toggle.bind(this);
    }

    toggle() {
      this.setState({isOpen: !this.state.isOpen});
    }

    render() {
      return(
        <div className='panel-wrapper m-0 p-1'>
          <div onClick={this.toggle}> { (this.state.isOpen ? '\u25be' : '\u25b8') + ' ' + title}</div>
          <Collapse className='mt-1 pl-2 pr-2' isOpen={this.state.isOpen}>
            <WrappedComponent {...this.props}/>
          </Collapse>
        </div>
      );
    }
  }
}

// ---- TableView ----
// Displays the module's rendered output, if any


class TableView extends React.Component {
  constructor(props) {
    super(props);
    this.state = { tableData: [], loading: false, isOpen: false };           // componentDidMount will trigger first load
    this.loadingState = { tableData: [], loading: true };
    this.toggle= this.toggle.bind(this);
  }

  // Load table data from render API
  loadTable() {
    var self = this;
    var url = '/api/wfmodules/' + this.props.id + '/render';
    fetch(url, { credentials: 'include'})
      .then(response => response.json())
      .then(json => {
        self.setState(Object.assign({}, this.state, {tableData: json, loading: false}));
      }); // triggers re-render
  }

  // Load table when first rendered
  componentDidMount() {
    this.loadTable()
  }

  // If the revision changes from under us reload the table, which will trigger a setState and re-render
  componentWillReceiveProps(nextProps) {
    //console.log("willRecieveProps " + this.props.id);
    //console.log('old revision ' + this.props.revision + ' new revision ' + nextProps.revision);

    if (this.props.revision != nextProps.revision) {
      this.setState(Object.assign({}, this.state, this.loadingState));               // "unload" the table
      this.loadTable();
    }
  }

  // Update only when we are not loading
  shouldComponentUpdate(nextProps, nextState) {
    return !nextState.loading;
  }

  toggle() {
    this.setState(Object.assign({}, this.state, { isOpen: !this.state.isOpen}));
  }

  render() {
    var tableData = this.state.tableData;
    var table;

    // Generate the table if there's any data
    if (tableData.length > 0 && !this.state.loading) {
      var columns = Object.keys(tableData[0]).filter(key => key!='index').map( key => { return { 'key': key, 'name': key, 'resizable':true } });
      table = <ReactDataGrid
        columns={columns}
        rowGetter={ i => tableData[i] }
        rowsCount={tableData.length}
        minHeight={500} />;
    }  else {
      table = <p>(no data)</p>;
    }

    return table;
  }
}



// ---- WfModule ----

//  Some convenient components that collapse params, output
const CollapsibleTableView = CollapseSection(
  TableView,
  'Output',
  true);     // don't start open

const ParamDivsComponent = (props) => <div>{props.paramDivs}</div>
const CollapsibleParams = CollapseSection(
  ParamDivsComponent,
  'Settings',
  true);         // start open

export default class WfModule extends React.Component {

  constructor(props) {
    super(props);
    this.initFields(props);
    this.setParamText = this.setParamText.bind(this);
    this.getParamText = this.getParamText.bind(this);
    this.removeModule = this.removeModule.bind(this);
  }

  componentWillReceiveProps(newProps) {
    this.initFields(newProps)
  }

  // our props are annoying (we use data- because Sortable puts all these props om the DOM object)
  // so save them into this
  initFields(props) {
    this.wf_module = props['data-wfmodule'];
    this.module = this.wf_module.module;
    this.params = this.wf_module.parameter_vals;
    this.revision = props['data-revision'];
  }

  // These functions allow parameters to access each others value (text params only)
  // Used e.g. for custom UI elements to save/restore their state from hidden parameters
  // Suppresses reassignment of the same text, which can be important to avoid endless notification loops
  setParamText(id_name, text) {
    var p = this.params.find( p => p.parameter_spec.id_name == id_name );
    if (p && text != p.text) {
      this.props['data-changeParam'](p.id, { text: text })
    }
  }

  getParamText(id_name) {
    //console.log("getParamText " + id_name)
    var p = this.params.find( p => p.parameter_spec.id_name == id_name );
    if (p) {
      return p.text;
    }
  }

  removeModule() {
    this.props['data-removeModule'](this.wf_module.id);
  }

  render() {

    // Each parameter gets a WfParameter
    var paramdivs = this.params.map((ps, i) => {
        return <WfParameter
          key={i}
          p={ps}
          changeParam={this.props['data-changeParam']}
          wf_module_id={this.wf_module.id}
          revision={this.revision}
          getParamText={this.getParamText}
          setParamText={this.setParamText} />
      });

    // Putting it all together: name, status, parameters, output
    return (
      <div className='container' {...this.props} >
        <div className='card w-75 mx-auto mb-4 bg-faded'>
          <div className='card-block drop-shadow p-1'>

            <div className='d-flex justify-content-between align-items-center mb-2'>
                <button type='button' className='btn btn-secondary btn-sm' onClick={this.removeModule}>&times;</button>
                <h3 className='text-center mb-0'>{this.module.name}</h3>
                <StatusLight status={this.wf_module.status}/>
            </div>
            <StatusLine status={this.wf_module.status} error_msg={this.wf_module.error_msg} />
            <CollapsibleParams paramDivs={paramdivs}/>
            <CollapsibleTableView id={this.wf_module.id} statusReady={this.wf_module.status == 'ready'} revision={this.revision} />
            <a className='ml-2' href={'/public/moduledata/live/' + this.wf_module.id + '.csv'}>CSV</a>/<a href={'/public/moduledata/live/' + this.wf_module.id + '.json'}>JSON</a>
          </div>

        </div>
      </div>
    ); 
  } 
}

