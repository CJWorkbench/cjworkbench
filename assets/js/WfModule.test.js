import React from 'react'
import { WfModule } from './WfModule'
import { okResponseMock } from './utils'
import { mount } from 'enzyme'

describe('WfModule, not read-only mode', () => {

  let wrapper;
  let props;
  let mockApi;

  // A basic mock module, set up to test the slightly gnarly press enter on URL field = press URL fetch button logic
  var wf_module = {
    'id' : 999,
    'notes': [],
    'parameter_vals': [
      {
        'id': 100,
        'parameter_spec': {
          'id_name': 'url',
          'type': 'string'
        }
      },
      {
        'id': 101,
        'parameter_spec': {
          'id_name': 'version_select',
          'type': 'custom'
        }
      }
    ],
    'module_version' : {
      'module' : {
        'id_name' : 'loadurl'
      },
    }
  };


  beforeEach(() => {
    // Reset mock functions before each test
    mockApi = {
      'postParamEvent' : okResponseMock(),
      'onParamChanged' : okResponseMock()
    };

    props = {
        'data-isReadOnly': false,
        'data-wfmodule': wf_module,
        'data-changeParam': mockApi.onParamChanged,
        'data-removeModule':  () => {} ,
        'data-revision': 707,
        'data-selected': true,
        'data-api': mockApi,
        'connectDragSource': jest.fn(),
        'connectDropTarget': jest.fn(),
        'focusModule': jest.fn(),
        'toggleDrag' : jest.fn()
      };

    wrapper = mount(
      <WfModule
        {...props}
      />
    )
  });

  it('Renders and calls functions on componentDidMount and render', () => {
    expect(wrapper).toMatchSnapshot();
    expect(props.connectDragSource.mock.calls.length).toBe(1);
    expect(props.connectDropTarget.mock.calls.length).toBe(1);
    expect(props.focusModule.mock.calls.length).toBe(1);
  });

  // Hardcoded logic where the url parameter communicates with the version select parameter
  it('pressing enter on a URL field fetches', (done) => {

    expect(wf_module.module_version.module.id_name).toBe('loadurl');
    let id = wf_module.parameter_vals[0].id;
    let paramIdName = wf_module.parameter_vals[0].parameter_spec.id_name;
    expect(paramIdName).toBe('url');
    let payload = { 'value' : 'foo' };

    let pressedEnter = false;
    wrapper.instance().changeParam(id, paramIdName, payload, pressedEnter);
    expect(mockApi.postParamEvent.mock.calls.length).toBe(0); // didn't press enter, no click event

    pressedEnter = true;
    wrapper.instance().changeParam(id, paramIdName, payload, pressedEnter);

    // setImmediate b/c changeParam waits for onParamChanged to complete before postParamEvent -- hence a promise
    setImmediate( () => {
      expect(mockApi.postParamEvent.mock.calls.length).toBe(1); // enter on url field, fetch data
      done();
    });
  });

});
