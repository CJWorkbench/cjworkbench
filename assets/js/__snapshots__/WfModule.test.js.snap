// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfModule, not read-only mode Renders and calls functions on componentDidMount and render 1`] = `
<WfModule
  connectDragSource={[Function]}
  connectDropTarget={[Function]}
  data-api={
    Object {
      "onParamChanged": [Function],
      "postParamEvent": [Function],
    }
  }
  data-changeParam={[Function]}
  data-isReadOnly={false}
  data-removeModule={[Function]}
  data-revision={707}
  data-selected={true}
  data-wfmodule={
    Object {
      "id": 999,
      "module_version": Object {
        "module": Object {
          "id_name": "loadurl",
        },
      },
      "notes": Array [],
      "parameter_vals": Array [
        Object {
          "id": 100,
          "parameter_spec": Object {
            "id_name": "url",
            "type": "string",
          },
        },
        Object {
          "id": 101,
          "parameter_spec": Object {
            "id_name": "version_select",
            "type": "custom",
          },
        },
      ],
    }
  }
  focusModule={[Function]}
  toggleDrag={[Function]}
>
  <div
    onClick={[Function]}
  >
    <div
      className="wf-card mx-auto"
    />
  </div>
</WfModule>
`;
